user/src/com/google/gwt/user/client/ui/RootPanel.java

```diff
@@ -89,11 +89,18 @@ public void execute(Widget w) {
    * This method may only be called per widget, and only for widgets that were
    * originally passed to {@link #detachOnWindowClose(Widget)}.
    * </p>
-   * 
+   * <p>
+   * Note that modern browsers do not have the memory leaks that originally required use of this
+   * feature - it is retained only to support application-specific detach events.
+   * </p>
+   *
    * @param widget the widget that no longer needs to be cleaned up when the
    *          page closes
    * @see #detachOnWindowClose(Widget)
+   * @deprecated Instead, prefer to use the {@code bfcache}, {@code pagehide},
+   * {@code visibilitychange} events, etc.
    */
+  @Deprecated
```

Comment:
These changes go too far as we should not make these methods deprecated unless we want `Widget.onDetach` to become public.

I have looked through our app and we use `onLoad` and `onUnload` to:
- register/deregister event handlers for widget events (DOM as well as logical widget events). This was for old browsers but could be skipped today.
- register/deregister events on an app wide singleton eventbus
- register/deregister with other singleton classes, e.g. an autosave manager which handles timers internally to fire autosave events for registered widgets
- start/stop timer for a given widget for various tasks, e.g. something simple as a clock.
- reset internal widget state if we want to reuse a widget, so it starts with fresh state on attach
- probably a couple of other things

Now imagine you have a GWT application that loads HTML from the server, renders the HTML using `setInnerHTML` and then uses the various `*.wrap(Element)` methods of GWT widgets as well as custom widgets to make that HTML interactive. If the application loads and unloads many such HTML fragments it is important to cleanup the widgets corresponding to a single HTML fragment using `detachNow`. Otherwise they will stay in the static widget list in `RootPanel` until you leave the page causing memory to increase and increase.
Now you could argue that these `wrap` widgets should not put themselves into the detach list anymore because they don't have to (seen from a DOM event perspective). But then `onUnload` lifecycle methods won't work anymore and it will cause trouble for most of the above use cases.

This change is only about what happens if we close the window and it should not interfere with widget lifecycle in general while the application runs normally and just loads/unloads elements/widgets.

So I do think we cannot get rid of `detachNow` and as a consequence `isInDetachList` and `detachOnWindowClose` (possibly renamed to `addToDetachList`) in the future unless we decide to make `Widget.onDetach` public as a non-deprecated replacement for `detachNow` to support widget lifecycle methods for `Widget.wrap(Element)` widgets. Instead of `detachNow` users would then need to call `onDetach` directly and might need to reimplement a detachlist themselves if they need `isInDetachList` for some reason.

Reply:
I think there's a compromise here that still lets us get rid of all of this (deprecating for now, probably leaving it all for at least a year, probably longer), but first let's clarify exactly what we're taking away and what it implies:

 * We're _not_ removing attach/detach or load/unload. At that point, we might as well get rid of Widget itself - definitely a choice that projects can make, but one we don't want to make for them.
 * Instead, the root here is the onclose handler that Window itself maintained, and provided for classes like RootPanel. This must be removed since otherwise we're actively preventing projects from using `bfcache`.
 * With that removed, RootPanel in turn must stop listening for the close, since likewise it would actively prevent projects from using `bfcache`. 
 * With RootPanel.detachOnClose gone, these other calls end up turning into a memory leak _especially_ in the case of `bfcache` being functional, so we want to actively discourage them - if a project is using one of the `wrap` factory methods with `detachNow`, they should switch instead to `widget.removeFromParent()`. This is not clear in the current version of the patch. This is strictly superior to exposing `onDetach` publicly, since it does checks to see how to detach the widget.
 * We definitely don't want projects to _start_ using `detachNow`, but existing projects can continue to use it and should expect it and `isInDetachList` to continue to behave properly.

Proposed migration path:
 * In GWT 2.N, we will deprecate these methods and improve documentation. Teams can continue to use these methods for some time, but documentation will encourage migrating from `detachNow(foo)` to `foo.removeFromParent()`. To continue to handle the memory leak in `RootPanel.widgetsToDetach`, `removeFromParent` will continue to call `detachNow(this)` as it has done. The patch already documents that calling `detachWidgets()` is required as the page closes if that "best effort" call is important to their app (but since it wasnt reliable to begin with, it shouldn't be assumed to work). 
 * In GWT 2.(N+1) or so, we will update Widget.removeFromParent() to stop calling detatchNow(this), and instead directly call onDetach - after an assert on `attached` to ensure that the widget should even be detached to begin with. The combined "is attached" and "parent is null" check should correctly account for the case of a widget that was created via `wrap()`, has not yet been detached, and may or may not still be in the dom. 
 * At this time, `detachNow` should be a no-op, and likewise `detachOnWindowClose` etc - the entire `widgetsToDetach` collection should go away, and calls to these methods should warn  - there's no memory leak, but detach isn't getting called any more here.
 * In GWT 2.(N+2), remove these deprecated no-op methods, to ensure swift migration and no surprises when things don't work.

I believe `RootPanel.get()` should be treated like `wrap()` calls through all of this - if you elect to call detachWidgets at some point, they will get cleaned up together. Alternatively, a RootPanel can be cleaned up directly through a call to removeFromParent().

Risks I see:
 * No one reads docs or release notes fully (obviously myself included). Some people check their build log when they see deprecations. Most people check runtime logs when bugs start happening. It is possible we should move the warnings to GWT 2.N so that we make noise a little sooner.
 * Projects that incorrectly call detachWidgets() while using the bfcache will be surprised that their events no longer work. Javadoc and release notes should at least point them in the right direction here.

With all of that said, attach/detach on page close will still do no better than best effort, but requires active participation in one place in a project after this patch - and it not called, these widgets should function correctly when using `bfcache` automatically, as they weren't detached when the page was hidden (and so you don't want unload/load going off). But when rearranging widgets or removing them from their `HasWidgets` parents, they will correctly continue to call `onAttach`/`onDetach` and fire their attach events.

---

If this clears up the issue, I have the meat of this implemented locally and a few new unit tests to confirm, but am still working on updating documentation.

Reply:
Pushed an update reflecting this, please take a look.

Reply:
Can we use `@deprecated(forRemoval = true)`? It was added in Java 9 and we compile using `--release 8`. But given it is not new syntax, it might work? Then CI tooling might create a little noise.

Reply:
GWT's own implementation of Override doesn't yet support forRemoval/etc. Maybe hit that in the next release, and we can sprinkle it around?

Reply:
Oh ok. Yes then some other day.